src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,134 @@
+        KalmanFilter<N1, N1, N1> kalman = new KalmanFilter<>(Nat.N1(), Nat.N1(), stateSpace,
+                VecBuilder.fill(MODEL_TOLERANCE),
+                VecBuilder.fill(ENCODER_TOLERANCE),
+                Constants.ROBOT_TIMEOUT
+        );
+        LinearQuadraticRegulator<N1, N1, N1> lqr = new LinearQuadraticRegulator<>(stateSpace, VecBuilder.fill(VELOCITY_TOLERANCE),
+                VecBuilder.fill(Constants.VOLTAGE),
+                Constants.ROBOT_TIMEOUT // time between loops, DON'T CHANGE
+        );
+        this.stateSpacePredictor = new LinearSystemLoop<>(stateSpace, lqr, kalman, Constants.VOLTAGE, Constants.ROBOT_TIMEOUT);
+    }
+
+    /**
+     * @return the velocity of the shooter. [rps]
+     * @see #setVelocity(double)
+     */
+    public double getVelocity() {
+        return rpsUnitModel.toVelocity(main.getSelectedSensorVelocity());
+    }
+
+    /**
+     * Set the velocity to apply by the motor.
+     *
+     * @param velocity the desired velocity at which the motor will rotate. [rps]
+     * @see #setPower(double)
+     */
+    public void setVelocity(double velocity) {
+        stateSpacePredictor.setNextR(VecBuilder.fill(velocity)); //r = reference
+        stateSpacePredictor.correct(VecBuilder.fill(getVelocity()));
+        stateSpacePredictor.predict(Constants.ROBOT_TIMEOUT); //every 20 ms
+
+        double voltageToApply = stateSpacePredictor.getU(0); // u = input, calculated by the input.
+        // returns the voltage to apply (between 0 and 12)
+        setPower(voltageToApply / Constants.VOLTAGE); // map to be between 0 and 1
+    }
+
+    /**
+     * Set the power to apply by the motor.
+     *
+     * @param power the power at which the motor will rotate. [percentage, between 0 and 1]
+     * @see #stop()
+     */
+    public void setPower(double power) {
+        main.set(TalonFXControlMode.PercentOutput, power, DemandType.ArbitraryFeedForward, Constants.Shooter.ARBITRARY_FEED_FORWARD);
+    }
+
+    /**
+     * Estimate the velocity that the shooter should apply in order to reach the target.
+     * The function created by running an experiment of different velocities for the "most used" distances.
+     * See the experiment results in (TODO: Add the path).
+     *
+     * @param distance the distance from the target.
+     * @return the velocity that should be applied by the shooter in order to reach the target.[rps]
+     */
+    public double estimateVelocity(double distance) {
+        return distance;
+    }
+
+    /**
+     * Get whether the flywheel has reached the desired velocity in order to reach the target.
+     * Also, this function checks whether the flywheel has enough velocity in order to move the motor in first place.
+     *
+     * @param desiredVelocity the desired velocity at the motor will rotate. [rps]
+     * @return whether the flywheel reaches the desired velocity.
+     */
+    public boolean isReady(double desiredVelocity) {
```

Comment:
Has the shooter done his homework and is now ready for a test? Or does it have a date and is currently looking for some nice clothes, so it's not ready yet?

I know you put documentation that explains what "ready" means, but it should be clear from the method's name, not from the documentation.
```suggestion
    public boolean hasReachedSetpoint(double desiredVelocity) {
```
Or:
```suggestion
    public boolean hasReachedVelocitySetpoint(double desiredVelocity) {
```
I think the first is better.

Reply:
I don't think we should use setpoint as a parameter because the word most of the time is used (most of the time) to describe a position, maybe `hasReachedVelocity`

Reply:
That's not true.
The definition of "setpoint" is "the desired or target value for an essential variable, or process value of a system" ([source](https://en.wikipedia.org/wiki/Setpoint_(control_system))). It could be a target position, but it can also be a certain velocity value or anything else you can feed your controller.

Reply:
so "setpoint" doesn't really clarify what the function does...
if it's only a "target value" it could be the velocity/position/acceleration/etc...

Reply:
That's why I suggested `hasReachedVelocitySetpoint`. But anyone who knows this subsystem's role will understand it's a velocity setpoint, not a position or an acceleration one.